# Basic `pandas`

In this notebook we will learn about `pandas` a vital data management package in python.

By the end of this notebook you will know about:
- `pandas` `Series` and `DataFrame` objects,
- Indexing `pandas` objects,
- Reading files with `pandas`,
- Saving files with `pandas` and 
- Some helpful `pandas` functionality.

## `pandas`

`pandas` is one of the most popular data handling packages in `python`. We will cover the minimum you need to know about the package for the boot camp in this notebook.

Let's start by importing the package. Just like in the previous notebook this should check whether or not you have `pandas` installed on your machine. If you are using Anaconda, <a href="https://www.anaconda.com/">https://www.anaconda.com/</a>, `pandas` should be installed already. If not check out the documentation for installation instructions <a href="https://pandas.pydata.org/docs/getting_started/install.html">https://pandas.pydata.org/docs/getting_started/install.html</a>.

In [1]:
## it is standard to import pandas as pd
import pandas as pd

In [2]:
## let's check what version of numpy you have
## when I wrote this I had version 1.5.3
## yours may be different
print(pd.__version__)


2.2.3


If you had a version of `pandas` installed, both of those code chunks should have executed without error. If not, you will need to install it onto your machine because we will be using it heavily in the boot camp. If you are unsure how to install a python package in general check our python package installation guide, <a href="https://www.erdosinstitute.org/data-science">https://www.erdosinstitute.org/data-science</a>.

##### Be sure you can run both of the above code chunks before continuing with this notebook, again it should be fine if your package version is slightly different than mine.

### `Series` and `DataFrame`s

`pandas` has two main data structures: 
1. `Series` objects, <a href="https://pandas.pydata.org/docs/reference/api/pandas.Series.html">https://pandas.pydata.org/docs/reference/api/pandas.Series.html</a> and 
2. `DataFrame` objects, <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html</a>. 

Let's explore them below.

In [3]:
## We can turn a list into a series
## with pd.Series()
list1 = [0,1,2,3]
print("Below is a list.")
print(list1, type(list1))
print()
print("Below is a pandas Series object.")
print(pd.Series(list1), type(pd.Series(list1)))

Below is a list.
[0, 1, 2, 3] <class 'list'>

Below is a pandas Series object.
0    0
1    1
2    2
3    3
dtype: int64 <class 'pandas.core.series.Series'>


The second thing we printed was a `Series` object. Note the two columns of numbers. The first column is the index of the object, the second column contains the values of the object. We can access those two separately like below.

In [4]:
series1 = pd.Series(list1)

In [5]:
## The index
series1.index

RangeIndex(start=0, stop=4, step=1)

In [6]:
## The values
series1.values

array([0, 1, 2, 3])

In [8]:
## You practice
## Take the list labeled a and 
## turn it into a Series named b
a = [5,2,3,6,'a','b','e',True,False]
b = pd.Series(a)
b


0        5
1        2
2        3
3        6
4        a
5        b
6        e
7     True
8    False
dtype: object

In [10]:
## We can get select entries from a Series 
## using the index
b[3]


6

In [11]:
## Note that we don't have to have our index be a list of numbers
c = pd.Series([1,2,3,4], index=['a','b','c','d'])

print(c)

a    1
b    2
c    3
d    4
dtype: int64


In [16]:
## and we can access by index in the same way
print(c['b'])

2


In [15]:
c.iloc[0]

np.int64(1)

Now let's check out a `DataFrame`. `DataFrame`s are essentially a collection of `Series` with a common index. We can also think of them as a table of values with column and row labels.

In [17]:
## We can make a DataFrame using a dictionary
## the dictionary keys are the column labels
## the dictionary values are columns
df = pd.DataFrame({'one':[3,4,5,2,4,5], 
                    'two':['a','b','e','h','l','p'],
                    'third column':[7,7,7,7,7,7]})

## NOTE that this is not the only way to make 
## a dataframe!

print(df)

   one two  third column
0    3   a             7
1    4   b             7
2    5   e             7
3    2   h             7
4    4   l             7
5    5   p             7


This is a `DataFrame`, the unlabeled column is the index, the labeled columns are `Series` objects themselves. We can access them in the following way.

In [18]:
## If your column's name doesn't violate a couple of format rules
## you can use
## df.column_name
print(df.one) 
print()
print("Note that each column is a Series object.")
print(type(df.one))

0    3
1    4
2    5
3    2
4    4
5    5
Name: one, dtype: int64

Note that each column is a Series object.
<class 'pandas.core.series.Series'>


In [ ]:
## However if our column name has spaces or certain characters in it
## (like . , ! ? "" and so on)
## we can't use .column_name


In [19]:
## So we have to use df["column name"] instead
# or df['column_name']
df['third column']

0    7
1    7
2    7
3    7
4    7
5    7
Name: third column, dtype: int64

In [20]:
## Just like with a Series we can use .index
## to get the index
df.index

RangeIndex(start=0, stop=6, step=1)

In [21]:
## .values returns a 2-D numpy ndarray with our columns
df.values

array([[3, 'a', 7],
       [4, 'b', 7],
       [5, 'e', 7],
       [2, 'h', 7],
       [4, 'l', 7],
       [5, 'p', 7]], dtype=object)

In [24]:
## You code
## Make a data frame, call it my_df 
## Make the first column labeled 'first' from a
## Make the second column labeled 'second' from b
## see what happens when you add , index=range(10,10+len(a)) 
## after the dictionary
a = [4,5,3,4,5,6,0]
b = ['a','c','d','g','l','m','p']

my_df = pd.DataFrame({'first':a, 'second':b}, index = range(10,10+len(a)))




In [25]:
my_df

first second
10      4      a
11      5      c
12      3      d
13      4      g
14      5      l
15      6      m
16      0      p

#### Locating `DataFrame` Entries With the Index

Locating entries in a `DataFrame` is slightly more complicated than what we had to do for `Series` objects.

In [28]:
## You code
## try to index the dataframe df like you would
## a list or Series to find the entry corresponding
## to index 1


## What happens?

In order to get particular `DataFrame` entries you have to use one of two methods, `.loc[]` or `.iloc[]`.

In [30]:
my_df.iloc[1]

first     5
second    c
Name: 11, dtype: object

In [ ]:
## for example we can get the 1 row of df like so with loc


In [ ]:
## or like so with iloc


The difference between `.loc` and `.iloc` is that `.iloc` is restricted to integer based indexing, while `.loc` is more versatile.

This means that if your index is a list of strings you cannot use `.iloc` with just the index's normal value. For example try to run the following.

In [ ]:
## Recall what c is


In [ ]:
## now try to get the ath entry using .iloc


In [ ]:
## instead we'd have to notice that the "a" entry is
## the 0 entry if the index were a normal integer based index
c.iloc[0]

In [ ]:
## However, we can just insert "a" into .loc
c.loc["a"]

Another difference is that `.loc` can be used to perform boolean indexing (eg. find me all the rows where the first column is between $2$ and $4$)

In [33]:
df

one two  third column
0    3   a             7
1    4   b             7
2    5   e             7
3    2   h             7
4    4   l             7
5    5   p             7

In [44]:
int(df.loc[3,'third column'])

7

In [34]:
## give me all the rows where the "one" column is between 2 and 4
df.loc[(df.one > 2) & (df.one < 4)]

one two  third column
0    3   a             7

In [37]:
## we can even go one step further and find a specific column
df.loc[(df.one > 1) & (df.one < 4), ['two','third column']]

two  third column
0   a             7
3   h             7

In [ ]:
## or a subset of many columns
## in this case you enter the subset as a list of column name strings
df.loc[]

In [ ]:
## order of columns can even be different from dataframe
df.loc[(df.one < 4) & (df.one > 2), ['third column','two']]

### Reading in a `.csv` file with `pandas`

You can also read in common data file types with `pandas`. Let's load the following.

In [38]:
## You can read in a csv with
## pd.read_csv("filename.csv")
## https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

## We'll work with the following dataframe in the next section
jr_shots = pd.read_csv("../data/JR_Smith_Shots_2015_16.csv")

In [39]:
jr_shots

LOC_X  LOC_Y  SHOT_MADE_FLAG
0     -106    244               0
1      -96     97               0
2       30     23               0
3     -204     -1               0
4      -76    237               0
..     ...    ...             ...
844    145    -15               0
845   -241     67               0
846    164    195               0
847      0      1               1
848    -89    288               0

[849 rows x 3 columns]

In [45]:
made = 0
for i in range(849):
    made += int(jr_shots.loc[i,'SHOT_MADE_FLAG'])
print('made =',made)
print('pct =', made/849)

made = 352
pct = 0.41460541813898705


In [40]:
## You code
## read in the file "beer.csv" using pandas
## call it beer_df
beer_df = pd.read_csv('../data/beer.csv')

## you'll practice with this dataframe in the next section

In [41]:
beer_df

IBU   ABV  Rating Beer_Type
0     45   4.2   3.792     Stout
1     60   8.3   4.145     Stout
2     25   6.0   3.951     Stout
3     31  11.0   4.062     Stout
4     75   9.0   4.018     Stout
..   ...   ...     ...       ...
342   68   8.2   3.922       IPA
343   45   6.8   4.000       IPA
344  101   9.2   3.778       IPA
345   70   6.8   4.015       IPA
346   80   6.9   3.554       IPA

[347 rows x 4 columns]

There are other useful `read` functions like `read_table` and `read_json`, you will see those later.

### Helpful `pandas` functions

`pandas` offers some really nice built in functions to help you explore any data set you're dealing with. Let's explore them below.

In [47]:
## df.head(n) let's you inspect the first n rows of the dataframe
## n defaults to 5
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html
jr_shots.head(11)

LOC_X  LOC_Y  SHOT_MADE_FLAG
0    -106    244               0
1     -96     97               0
2      30     23               0
3    -204     -1               0
4     -76    237               0
5      25     23               1
6      43     47               1
7      48    100               0
8      22     16               1
9      19     16               0
10    163    141               0

In [49]:
## You code 
## investigate what .tail(n) does 
## using the beer_df
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.tail.html
## Hint: n must be an integer
beer_df.tail(11)

IBU  ABV  Rating Beer_Type
336   70  9.2   3.846       IPA
337  100  9.6   3.978       IPA
338   45  7.0   3.395       IPA
339   54  6.3   3.679       IPA
340   38  6.9   3.647       IPA
341   70  7.2   3.549       IPA
342   68  8.2   3.922       IPA
343   45  6.8   4.000       IPA
344  101  9.2   3.778       IPA
345   70  6.8   4.015       IPA
346   80  6.9   3.554       IPA

In [52]:
## You code
## investigate what .sample(n) does
## using the beer_df
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sample.html
## Hint: n must be an integer
beer_df.sample(12)

IBU   ABV  Rating Beer_Type
72    38   6.5   4.042     Stout
264   69   6.5   3.617       IPA
272   80   6.4   3.671       IPA
322   50   6.6   3.982       IPA
219   30   5.4   3.476       IPA
185   65   6.8   3.737       IPA
320   60   7.0   3.743       IPA
324   65   7.0   3.656       IPA
70    50  10.7   4.068     Stout
280   90  10.0   3.513       IPA
2     25   6.0   3.951     Stout
68    45  10.2   3.919     Stout

In [55]:
## You code
## what happens when you add the argument
## random_state=440 in .sample()?
## rerun the code chunk multiple times with or without the random_state
## try different numbers for the random_state
beer_df.sample(11, random_state = 440)

IBU   ABV  Rating Beer_Type
301   62   6.8   4.073       IPA
5     60  15.2   4.404     Stout
182   82   7.3   3.738       IPA
143   55   8.1   3.776     Stout
4     75   9.0   4.018     Stout
112   70  12.2   4.375     Stout
316   35   6.8   3.701       IPA
46     6   9.0   3.793     Stout
202   75   7.2   3.800       IPA
89    25   9.0   3.564     Stout
116   70  10.0   4.004     Stout

In [56]:
## df.info() tells you useful information about your
## dataframe, like the column names,
## the number of rows,
## the number of non-empty entries for each column,
## the data type of each column
## and how much memory the dataframe uses
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html
jr_shots.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 849 entries, 0 to 848
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   LOC_X           849 non-null    int64
 1   LOC_Y           849 non-null    int64
 2   SHOT_MADE_FLAG  849 non-null    int64
dtypes: int64(3)
memory usage: 20.0 KB


In [57]:
## df.describe()
## provides the 5 stat breakdown of each numeric column
## that is any column that consists of integers or floats
## will have its, minimum, maximum, first quartile, median,
## third quartile, and mean provided
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html
jr_shots.describe()

LOC_X       LOC_Y  SHOT_MADE_FLAG
count  849.000000  849.000000      849.000000
mean    -6.204947  112.106007        0.414605
std    157.201943   91.254429        0.492944
min   -250.000000  -46.000000        0.000000
25%   -158.000000   21.000000        0.000000
50%     -2.000000  110.000000        0.000000
75%    133.000000  193.000000        1.000000
max    245.000000  693.000000        1.000000

In [60]:
## You code
## Try out .mean() and .max() and .min()
## on beer_df
## Compare the results to .describe()'s output

## .mean() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html
## .max() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.max.html
## .min() - https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.min.html

beer_df.describe()

IBU         ABV      Rating
count  347.000000  347.000000  347.000000
mean    57.579251    7.639769    3.853980
std     24.100858    2.216800    0.228915
min      6.000000    3.800000    3.022000
25%     40.000000    6.200000    3.700500
50%     60.000000    7.000000    3.823000
75%     70.000000    8.850000    3.982500
max    250.000000   18.000000    4.541000

In [61]:
## What about variables that aren't numeric? Like
## classes
## You can use df.value_counts() for those
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.value_counts.html
jr_shots.SHOT_MADE_FLAG.value_counts()

SHOT_MADE_FLAG
0    497
1    352
Name: count, dtype: int64

The above tells us that 497 observations have a `SHOT_MADE_FLAG` of 0 and 352 have a `SHOT_MADE_FLAG` of 1.

In [63]:
## You code
## How many of each "Beer_Type" are there in beer_df?


## What happens when you input the argument normalize=True into value_counts()?

beer_df.Beer_Type.value_counts(normalize=True)

Beer_Type
IPA      0.56196
Stout    0.43804
Name: proportion, dtype: float64

One task that we will do quite often in the boot camp is split one data set into two non-overlapping data sets. Let's see how we can do that here.

In [65]:
## Let's first randomly sample 100 observations from jr_shots
## Note the .copy()
## this creates a deep copy of the DataFrame 
jr_1 = jr_shots.sample(100, random_state = 54888).copy()

In [66]:
jr_1

LOC_X  LOC_Y  SHOT_MADE_FLAG
302    227     -1               0
474    -81    174               0
647   -241    111               0
109     15     65               1
376     60     21               1
..     ...    ...             ...
149   -165    198               0
766    128    188               0
792    140    208               0
336     25      8               0
342    -40    180               1

[100 rows x 3 columns]

In [67]:
## To create jr_2 we'll just drop the indices of jr_1 from
## jr_shots using .drop()
## https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html

## Again we include .copy() to make a deep copy of the dataframe
jr_2 = jr_shots.drop(jr_1.index).copy()

In [68]:
jr_1

LOC_X  LOC_Y  SHOT_MADE_FLAG
302    227     -1               0
474    -81    174               0
647   -241    111               0
109     15     65               1
376     60     21               1
..     ...    ...             ...
149   -165    198               0
766    128    188               0
792    140    208               0
336     25      8               0
342    -40    180               1

[100 rows x 3 columns]

In [69]:
jr_2

LOC_X  LOC_Y  SHOT_MADE_FLAG
0     -106    244               0
1      -96     97               0
2       30     23               0
3     -204     -1               0
4      -76    237               0
..     ...    ...             ...
842    171    178               1
844    145    -15               0
845   -241     67               0
846    164    195               0
848    -89    288               0

[749 rows x 3 columns]

In [70]:
## Let's check that it worked like we thought it would
len(jr_1) + len(jr_2) == len(jr_shots)

True

In [75]:
## You code
## split beer_df into beer_1 and beer_2
## beer_1 should be a random sample of 150 observations
## beer_2 should be all rows not in beer_1
beer_1 = beer_df.sample(150).copy()
beer_2 = beer_df.drop(beer_1.index).copy()


In [76]:
beer_1.index

Index([ 96,  55, 182,  90, 188, 278,  82,  42, 152, 186,
       ...
       128, 227,  71,  23, 276, 214,  77, 305, 204,  16],
      dtype='int64', length=150)

In [78]:
len(beer_1) + len(beer_2) == len(beer_df)

True

### Saving a `DataFrame` to File

Just like we easily read in a data file we can use `pandas` to quickly save data to a file.

This is done with the function `df.to_csv()`, <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html">https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html</a>.

In [79]:
## Let's first make a data frame
df = pd.DataFrame({'a':[1,2,3],
                   'b':[2,4,6],
                   'c':[17,34,51]})

df

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

In [81]:
## Now we call df.to_csv(file_name)
df.to_csv('../data/our_first_dataframe.csv')

Go ahead and check your repository, you should now see `our_first_dataframe.csv` in there.

In [83]:
## You code
## read in "our_first_dataframe.csv" using read_csv
## then look at the df

df = pd.read_csv('../data/our_first_dataframe.csv')
df
## is anything off?

Unnamed: 0  a  b   c
0           0  1  2  17
1           1  2  4  34
2           2  3  6  51

When you ran `to_csv()` the default is to record the index as well. That is why your `DataFrame` has an unlabeled column when you read it back in using `read_csv`. If we wanted to avoid writing the index to file you include the argument `index=False`. 

In [84]:
## Make the DataFrame again
df = pd.DataFrame({'a':[1,2,3],
                   'b':[2,4,6],
                   'c':[17,34,51]})

df

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

In [85]:
## write it to file without the index
df.to_csv('../data/our_first_dataframe.csv',index = False)

In [87]:
## You code
## read in "our_first_dataframe.csv" using read_csv
## then look at the df again
df = pd.read_csv('../data/our_first_dataframe.csv')

df

a  b   c
0  1  2  17
1  2  4  34
2  3  6  51

That's all you will need to know from `pandas` to have a firm footing for our boot camp. We will learn more stuff as we go along, but when that time comes you should be ready!

--------------------------

This notebook was written for the Erd&#337;s Institute C&#337;de Data Science Boot Camp by Matthew Osborne, Ph. D., 2023.

Any potential redistributors must seek and receive permission from Matthew Tyler Osborne, Ph.D. prior to redistribution. Redistribution of the material contained in this repository is conditional on acknowledgement of Matthew Tyler Osborne, Ph.D.'s original authorship and sponsorship of the Erdős Institute as subject to the license (see License.md)